#84 -> work area 1: Read All endpoints

pospi opened the pull request:
Initial cleanup pass for #221; refactors list indexing to place responsibility of "list all" data structures in index zomes alongside other queries.

Just wanted to put this draft up for @weswalla & @Connoropolous benefit as we move into the read API portion of the MMR work, since this is a significant architectural refactor.

pospi committed:
update core libs to move responsibility of handling 'all' query APIs to indexing zomes

misconfiguration of an indexing zome will now lead to a zome call error in create_entry_identity

pospi committed:
remove old get_all_ API endpoints from EconomicEvent & EconomicResource CRUD zomes

pospi committed:
fixes to indexing zome API macros

add new error variant to single out indexing config related errors

pospi committed:
add ProcessSpecification index zome for specification DNA, now that 'list all' API responsibility has been moved to index zomes

pospi committed:
update lockfile

pospi committed:
wire up DNAs for specification DNA to standard index pattern

pospi committed:
wire up ProcessSpecification index zome & DNA config to CRUD zome

pospi committed:
wire up all CRUD zomes to delegate record registration requests to companion indexing zomes

pospi committed:
define unit index zome to store list API data structures

pospi committed:
update lockfile

pospi committed:
add missing unit_index zome in specification DNA

pospi committed:
fix formatting / wiring of API methods between controller & index zomes

pospi committed:
add an into_cursor() method on all Response types, for integrating into pagination helpers

pospi committed:
change all entry type names to mirror zome names, now that they are used to derive cell entrypoints in the associated indexing zome

pospi committed:
fix missing entryDefs for ProcessSpecification & Unit index zomes

pospi committed:
rebind EconomicEvent & EconomicResource list API endpoints

pospi committed:
fix return type handling & show upstream errors in cross-zome entry identity handling

pospi committed:
disable creating metadata link from identity to entry for now, also does not work as such under new split-zome list API architecture

pospi committed:
optimise iteration of edges in list output handler, re-use edge_cursors better in output

pospi committed:
fix missing EntryDefs in ResourceSpecification index zome

pospi commented:
Ok, looks good now. Had some lingering bugs to track down but looks good after most recent commits. "List all" functionality is now fully the responsibility of index zomes, rather than CRUD zomes.

Connoropolous committed:
Merge branch 'sprout' into feature/221-queryall-index-cleanup

Connoropolous commented:
I merged sprout branch in, so that CI will run here now 👍

Connoropolous commented:
@pospi there is one assertion failing on ubuntu in CI, does that also occur for you locally?

pospi committed:
update method name

Connoropolous commented on lib/hdk_records/src/record_helpers.rs:
```diff
@@ -164,13 +168,17 @@ pub fn create_record<I, R: Clone, B, C, E, S>(
     // link the identifier to the actual entry
     // :TODO: this isn't needed for reading anymore, but might be worthwhile retaining for legibility in the DHT. Needs consideration as to DHT size bloat tradeoff.
-    create_link(identity_address, entry_hash, LinkTag::new(crate::identifiers::RECORD_INITIAL_ENTRY_LINK_TAG))?;
+    // create_link(identity_address, entry_hash, LinkTag::new(crate::identifiers::RECORD_INITIAL_ENTRY_LINK_TAG))?;
```
what's the reason for commenting it out, vs deleting it?

Connoropolous reviewed (commented):
I have scanned the code, but most importantly at the moment the integration tests suggest that the response for the read_all_* queries are coming back empty, so I assume things aren't quite fully functionally integrated yet

pospi committed:
fix bug in list API read requests due to double-escaping of zome query function name

pospi commented:
The core of this issue is now fixed- the remaining errors in that test are due to list APIs returning duplicate results.

However since we are pursuing #220 I wonder whether it is a better course of action to fold that work into this PR and have the switchover of DHT indexing structures resolve the duplication issue as a side-effect?

Connoropolous commented:
@pospi cool. Where do the duplicate results come from?

pospi commented:
(for completeness) duplicate link writes between the indexing entries.

pospi committed:
Merge branch 'sprout' into feature/221-queryall-index-cleanup

pospi committed:
update Plan create API to use index zome config for registering new record

pospi committed:
add Cursor conversion method for Plan record data

pospi committed:
integrate hc_time_index errors with hdk_records as indexing errors

pospi committed:
Merge remote-tracking branch 'origin/sprout' into feature/221-queryall-index-cleanup

pospi committed:
Merge branch 'update-hdk' into feature/221-queryall-index-cleanup

pospi committed:
relax HDK ver dependencies to minver

pospi committed:
update hdk_time_index package to loosen HDK dependency

pospi committed:
update lockfile

pospi committed:
update time index lib for compatibility with latest HDK

pospi committed:
fix borrow error

pospi committed:
update trait bounds & link creation API calls per new HDK

pospi committed:
update lockfile

pospi committed:
update trait bounds for latest HDK

pospi committed:
implement a wrapped struct for every *Address type which implements a Timed* variant for attaching timestamp data upon entry creation, and RPC params for transferring

pospi committed:
pass timestamp to indexing zome when creating entry indexes

pospi committed:
wire up logic for inserting new entries into date-based indexes

pospi committed:
finish passing timestamp to indexing zome entry registration helper, add placeholder params for dates in list API

pospi committed:
add trait bound needed to use hc_time_index to pull record anchors from DHT

pospi committed:
switch API to use read identifiers to request remote records

pospi committed:
remove cruft

pospi committed:
expose convenience entry retrieval methods to other crates

pospi committed:
pass DateTime<Utc> to index creation methods and separate indexable data from storage data since storage does not contain timestamp info in this case

pospi committed:
remove need for IndexableEntry trait from index management methods and create index pointers without date-indexing

must use get_links_for_time_span instead of get_links_and_load_for_time_span and manually fetch entries since they do not contain timestamping info

pospi committed:
import new Timed* structs to RPC crates for use in indexing zomes

pospi committed:
update lockfile

pospi committed:
Merge branch 'sprout' into feature/221-queryall-index-cleanup

pospi committed:
Revert "import new Timed* structs to RPC crates for use in indexing zomes"

This reverts commit 8aa16ccd3b5d9b33072890090f4553a3b18b8fbb.

pospi committed:
remove refs and structs based on hc_time_index module in preparation for bespoke impl

pospi committed:
update lockfile

pospi committed:
remove another uneeded crate ref

pospi committed:
implement first pass alternative approach to writing time-ordered indexing structures & wire up to indexing zomes

pospi committed:
add sane defaults to DNA properties for index bucketing (30 second chunks)

pospi committed:
remove unneeded structs

pospi committed:
define TimeIndex entry def for all indexing zomes

pospi committed:
update lockfile

pospi committed:
add debug output for admin websocket URI so that it can be connected to with Playground & other tools

Connoropolous commented on lib/hdk_time_indexing/src/lib.rs:
```diff
@@ -0,0 +1,86 @@
+}
+
+// Parse configuration & setup library constants
+lazy_static! {
```
I don't know if it matters much, but instead of lazy_static I know that the holochain team uses OnceCell now, and it looks like it's likely to get moved into `std` too

pospi replied:
ah, well that's good to know. Advice might be the outcome of https://github.com/holochain/holochain/issues/1444 then

Connoropolous commented on lib/hdk_time_indexing/src/writing.rs:
```diff
@@ -0,0 +1,97 @@
+            origin_hash.to_owned(),
+            HdkLinkType::Any,
+            link_tag,
+        ).map_err(|_e| { TimeIndexingError::EntryNotFound(origin_hash.to_owned()) })?;
```
why this line? I'm not sure it reflects real holochain behaviour, does it?

pospi replied:
you mean the error mapping? (sorry, it's not clear exactly which line you mean in the mobile Github UI :S)

Connoropolous replied:
The error mapping yeah

pospi replied:
Gotcha. Well, "real" behaviour- probably not?
I'm moving to a pattern where records have to be committed to the local DHT before any indexes can be created for them. Calls to `create_link` fail if the entry being linked to doesn't exist, and I wanted to catch this as the library's own error type so that the reason for erroring is a bit clearer to the caller. If you check `append_to_time_index` you'll see that it ensures the pointer entry being linked to exists before populating the index data structures.

Connoropolous commented on lib/hdk_semantic_indexes/zome/src/lib.rs:
```diff
@@ -100,6 +105,50 @@ pub fn query_index<'a, T, O, C, F, A, S, I, J, E>(
     Ok(entries)
 }
 
+/// Given a type of entry, returns a Vec of *all* records of that entry registered
```
won't be "all" if the caller constrains the request by time frame

pospi replied:
will update comment header when I update the method (:

Connoropolous reviewed (commented):
this is just my first read through. I was moderate thorough. I haven't done anything hands on

Commit:
Merge branch 'sprout' into feature/221-queryall-index-cleanup

Connoropolous reviewed (commented):
I tried to check out this branch and run it but there were a few Rust compilation failures. I could figure out three of them, but not why `get_links_for_time_span` is being called but is entirely undefined in the codebase. I'd happily check out this branch again and do some looking into it if I can have it compile.

pospi committed:
wire up new agent zome creation method to indexing & cursor APIs

pospi committed:
disable method until working

pospi commented:
My bad, that method is being updated to not target the `hc_time_index` module and I hadn't pushed. Have just put through a commit to comment it out until the logic at least compiles.

Connoropolous commented:
@pospi I don't have additional time today past my solving for the logging issue, which I've accomplished (see #305 ) but I did also push a fix on this branch for an issue where if a test fails, `hc sandbox` could leave spawned holochain and lair-keystore processes running orphaned on the machine.

pospi committed:
WIP partial implementation of time-ordered index read logic

pospi commented:
Ah, great! Hectic change, thanks for taking the time (:

`ps aux | grep holochain` found like 12 of those I still had running!

pospi committed:
fix bug in leafmost chunk timestamp generation

I hate math.

pospi committed:
implement conversion method from IndexSegment to DateTime. Useful in debugging.

pospi committed:
create a static var to determine whether chunk leaves exist in the tree or not, so that querying knows correct depth

pospi committed:
move decoding logic to conversion method from LinkTag to IndexSegment

pospi committed:
WIP of index backward recursion and reading, needs lots of work

pospi committed:
implement outer methods for querying time indexes and wire up to index zomes macro

pospi committed:
add unit test runner script to workaround issues with compiling Holochain mocks for testing

pospi committed:
stub out (compiling) unit test crate for hdk_time_indexing

pospi committed:
ignore test lockfiles

Commit:
set up a first unit test with MockHDK

pospi committed:
move test file to sub-module to better separate

pospi committed:
start of tests for asserting written index paths (currently failing)

pospi committed:
abstract away existing mocking details into macros

pospi committed:
WIP work on mocking gets for PathEntry

Commit:
Merge branch 'sprout' into feature/221-queryall-index-cleanup

Commit:
make adjustments for agent zome post-merge

Commit:
other small merge adjustments for agent

pospi committed:
Merge branch 'sprout' into feature/221-queryall-index-cleanup

pospi committed:
pass arguments through to readAll endpoints

pospi committed:
remove temporary sorting, add pagination params to event/resource list API tests

pospi committed:
split up time index reading methods for re-use

pospi committed:
fix double escaping of index names

pospi committed:
implement simplest depth-first full traversal to read complete data from ordered indexes

pospi committed:
fix link ordering in index entry linkage wire-up

pospi committed:
reverse order of results in list API tests, since records are returned in FILO order

pospi committed:
update NPM lockfile

pospi commented:
First pass of this seems to be working! There is now a simple "complete traversal" method that is passing tests for the event & resource list APIs.
Ready for review- we can build up the more complex algorithms from this foundation. I am keeping a list of TODO items separately that I will probably log in a separate repo for the new `hdk_time_indexing` crate. Once we are happy with an MVP I think it will be best to split it out into its own repo and seek some systems programmers to take over its continuation :slightly_smiling_face:

Commit:
fix vf_plan to just plan

Connoropolous commented on lib/hdk_semantic_indexes/zome/src/lib.rs:
```diff
@@ -100,6 +110,48 @@ pub fn query_index<'a, T, O, C, F, A, S, I, J, E>(
+        SerializedBytes: TryInto<C, Error = SerializedBytesError> + TryInto<B, Error = SerializedBytesError>,
+        F: Fn(C) -> Option<String>,
+{
+    let linked_records = read_all_entry_hashes(index_name)
```
```suggestion
    // this algorithm is the 'make it work' current pass, pending the full implementation mentioned
    // in the TODO below, regarding efficiency and completeness
    let linked_records = read_all_entry_hashes(index_name)
```

Connoropolous reviewed (approved):
Fantastic. I see that you also added the missing indexes, which means that the only thing needed to make `processes`, `agents`, etc resolvers work is to wire up the resolvers, since you've moved the `read_all_x` endpoints into the index zomes. I ran the integration tests for the branch and they all passed, after making one small fix to from `vf_plan` to `plan` which was just a delayed merge thing

Commit:
Update lib/hdk_semantic_indexes/zome/src/lib.rs

Commit:
add inventoriedEconomicResources to DEFAULT_QUERIES in graphql explorer

Commit:
Merge branch 'feature/221-queryall-index-cleanup' of github.com:h-REA/hREA into feature/221-queryall-index-cleanup

Commit:
add a typescript type for the PagingParams
